### ISSUE TITLE
Calling a generic function with a generic class directly instantiated as its argument leads to wrong result type

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#4973]
**Describe the bug**
As the title.

**To Reproduce**
See sample code.

**Expected behavior**
Sample code should type-check.

**Screenshots or Code**
```python
from __future__ import annotations

from typing import Any, TypeVar
from typing_extensions import assert_type

T = TypeVar("T")

any: Any = globals()[""]

def f(x: list[T], /) -> list[T]:
    return x

a = assert_type(list([any]), "list[Any]")
b = assert_type(f(a), "list[Any]")  # passed

c = assert_type(f(list([any])), "list[Any]")  # failed
```
![image](https://user-images.githubusercontent.com/15389744/233020322-c9b99444-87d8-45a9-bec9-2aec210e6c37.png)

**VS Code extension or command-line**
Pylance v2023.4.21 Pre-Release

**Additional context**
- Mypy playground: https://mypy-play.net/?mypy=latest&python=3.10&gist=0474e7561600c1bc4098c30c7b37f0b2
- `list` is used for a quick repro. Any generic class can be used in place of `list`.
- Using `Any` and `Never` as type-var is triggering this behavior. Other concrete types seems not affected.
- I tend to remember there was a similar issue about this behavior, but haven't found it either in the Pyright repo or the Pylance repo.

**Comment by GugelRobin** [comment]
I'm having a similar issue were a TypeVar isn't resolved when instantiating a generic object in a constructor call.
Pinning the version to 1.1.304 mitigated the issue for us for now.

**Comment by erictraut** [comment]
@GugelRobin, if you're seeing a regression from 1.1.304 to 1.1.305, it's unlikely to be the same issue as this one. If you have a self-contained minimal repro, please file a separate issue.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.308, which I just published. It will also be included in a future release of pylance.
